tests/end_to_end/test_vulnerability_detector/test_vulnerability_detector.py

```diff
@@ -0,0 +1,863 @@
+    time.sleep(15)
+
+    # Start agents
+    host_manager.control_environment('start', ['agent'])
+
+    yield test_timestamp
+
+    # Truncate alerts and logs of managers and agents
+    logger.error("Truncate managers and agents logs")
+    truncate_remote_host_group_files(host_manager, 'all', 'logs')
+
+    logger.error("Restoring original configuration")
+    restore_configuration(host_manager, hosts_configuration_backup)
+
+    logger.error("Restarting environment")
+    host_manager.control_environment('restart', ['agent'])
+    host_manager.control_environment('restart', ['manager'])
+
+
+def check_vuln_state_consistency(vulnerabilities_alerts, vulnerabilities_states):
+    # Get the indexer values
+    alerts_vulnerabilities = []
+    indices_vulnerabilities = []
+
+    for alert in list(vulnerabilities_alerts.values())[0]:
+        alert_agent = alert['_source']['agent']['name']
+        alert_cve = alert['_source']['data']['vulnerability']['cve']
+        alert_package_version = alert['_source']['data']['vulnerability']['package']['version']
+        alert_package_name = alert['_source']['data']['vulnerability']['package']['name']
+        alerts_vulnerabilities.append({
+            'cve': alert_cve,
+            'agent': alert_agent,
+            'package_name': alert_package_name,
+            'package_version': alert_package_version
+        })
+
+    for vulnerabilities_state in list(vulnerabilities_states.values())[0]:
+        state_agent = vulnerabilities_state['_source']['agent']['name']
+        state_cve = vulnerabilities_state['_source']['vulnerability']['id']
+        state_package_name = vulnerabilities_state['_source']['package']['name']
+        state_package_version = vulnerabilities_state['_source']['package']['version']
+
+        indices_vulnerabilities.append({
+            'cve': state_cve,
+            'agent': state_agent,
+            'package_name': state_package_name,
+            'package_version': state_package_version
+        })
+
+    if len(alerts_vulnerabilities) != len(indices_vulnerabilities):
+        logger.critical("The number of alerts is not the same as the number of states")
+        logger.critical(f"Alerts: {len(alerts_vulnerabilities)}")
+        logger.critical(f"States: {len(indices_vulnerabilities)}")
+
+    alerts_not_in_states = []
+    states_not_in_alerts = []
+
+    # Check that all alerts are in the index
+    for alert in alerts_vulnerabilities:
+        if alert not in indices_vulnerabilities:
+            alerts_not_in_states.append(alert)
+
+    # Check that all index states are in the alerts
+    for state in indices_vulnerabilities:
+        if state not in alerts_vulnerabilities:
+            states_not_in_alerts.append(state)
+
+    return {
+        'alerts_not_in_states': alerts_not_in_states,
+        'states_not_in_alerts': states_not_in_alerts
+        }
+
+
+@pytest.mark.filterwarnings('ignore::urllib3.exceptions.InsecureRequestWarning')
+class TestInitialScans():
```

Comment:
Why is a class necessary?

Reply:
Tests classes are the approach followed in this case for the following reasons:
- Dependency as we required is not possible due to the complexity of the tests by default pytest modules. We need to track when and where a stage fails. Due to the current limitations of the framework for now, the best approach is to include all dependent tests into a class and handle the global results from there.
- Evidence gathering is a key part of testing. We need to create a global and standard data structure to handle this in the pytest html hooks, in order to include custom evidences to the tests. For now this is done by the get_results class fixture.